### PR TITLE
Rocket fix

### PR DESCRIPTION
**Pull request by simon33-2**
## Overview
Refer #4289.
Fixes a few problems with Rockets including #3846.
In the older games rockets will work as at present but for Global 1940 and derivatives, Rockets will now be considered after scrambling in all cases. Normally there will be rocket targeting, strategic bombing, rocket firing but there will be a map option to have strategic bombing happening before rocket targeting to support the unfortunate rule clarification.

## Functional Changes
See above.

## Manual Testing Performed
- ..

## Additional Review Notes

Reproducing my answers from #4289.
Re: ROCKET_ATTACKS_PER_FACTORY_INFINITE

At one point the property was used but somewhere along the line this function was moved to TechAbilityAttachment.

Re: Why was the default changed of RocketNumberPerTerritory to 1
Setting to 0 just caused rockets not to work. It may have been overridden by the INFINITE property not 100% sure now. In any event it didn't make much sense to have a non functional attribute and a property which did the same thing.

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java [hunk 1]
```diff
@@ -137,15 +137,17 @@ public static boolean isNonCombatMove(final GameData data, final boolean doNotTh
...
+   * Ideally, all maps with rockets will set PROPERY_fireRockets for move and battle phases.
    */
   static boolean isFireRockets(final GameData data) {
     data.acquireReadLock();
     try {
       final String prop = data.getSequence().getStep().getProperties().getProperty(GameStep.PropertyKeys.FIRE_ROCKETS);
       if (prop != null) {
         return Boolean.parseBoolean(prop);
+      } else if (data.getSequence().getStep().getDelegate().getName().equals("battle")) {
+        return Properties.getWW2V2(data) || Properties.getWW2V3(data);
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java]: That's correct.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 2]
```diff
@@ -68,64 +120,93 @@ private static boolean isLimitRocketDamageToProduction(final GameData data) {
...
+   *  Find Rocket Targets and load up fire rockets for later execution if necessary. Directly fired with strict rockets.
+   */
+  private void findRocketTargets(final IDelegateBridge bridge) {
     final GameData data = bridge.getData();
-    final Set<Territory> attackedTerritories = new HashSet<>();
-    final boolean oneAttackPerTerritory = !isRocketAttacksPerFactoryInfinite(data);
-    for (final Territory territory : rocketTerritories) {
-      final Set<Territory> targets = getTargetsWithinRange(territory, data, player);
-      if (oneAttackPerTerritory) {
-        targets.removeAll(attackedTerritories);
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Because we've made the max number of territory count in TechAbilityAttachment functional.

File: game-core/src/main/java/games/strategy/triplea/Constants.java [hunk 3]
```diff
@@ -76,7 +77,6 @@
   String ROLL_AA_INDIVIDUALLY = "Roll AA Individually";
   String LIMIT_ROCKET_AND_SBR_DAMAGE_TO_PRODUCTION = "Limit SBR Damage To Factory Production";
   String SBR_VICTORY_POINTS = "SBR Victory Points";
-  String ROCKET_ATTACKS_PER_FACTORY_INFINITE = "Rocket Attacks Per Factory Infinite";
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/Constants.java]: There is no map in the repo which has this property. Or is there somewhere else that it might be?

File: game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java [hunk 4]
```diff
@@ -137,15 +137,17 @@ public static boolean isNonCombatMove(final GameData data, final boolean doNotTh
...
       } else if (Properties.getWW2V2(data) || Properties.getWW2V3(data)) {
         return isCombatDelegate(data);
       }
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java]: Great question.

It seems that this function is never entered in combat move.

There may be a better way of doing this.

**Comment by simon33-2** [comment]
@Ron.

One thing I hate about this bit of code is referring to WW2V2 and/or WW2V3 properties. Is that because of the theory that V1 should be the default state?

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java [hunk 5]
```diff
@@ -189,13 +189,10 @@ public void end() {
...
-    // WW2V1, fires at end of non combat move
-    if (GameStepPropertiesHelper.isFireRockets(data)) {
-      if (needToDoRockets && TechTracker.hasRocket(bridge.getPlayerId())) {
-        RocketsFireHelper.fireRockets(bridge, bridge.getPlayerId());
-        needToDoRockets = false;
-      }
+    // WW2V1, fires at end of non combat move.
+    if (needToDoRockets && GameStepPropertiesHelper.isNonCombatMove(data, true)) {
+      RocketsFireHelper.fireWW2V1(bridge);
+      needToDoRockets = false;
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Tracking this through, it seems that the distance is set to zero if you don't the rocket tech. Not ideal. I've added a filter.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 6]
```diff
@@ -35,7 +38,60 @@
...
+    return object;
+  }
+
+  /**
+   * Fire rockets if we are v1
+   * In this rule set, each player only gets one rocket attack per turn. 
+   */
+  public static void fireWW2V1(final IDelegateBridge bridge) {
+    final GameData data = bridge.getData();
+    if (isWW2V2(data) || isAllRocketsAttack(data)) {
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Just moved it back to MoveDelegate but I think I see the reason. Putting it here has all the rocket code in the one place. I'm inclined to agree that it is awkward here. 

If we do keep it here, we probably should change the method to something like fireWW2V1IfNeeded

File: game-core/src/main/java/games/strategy/triplea/Constants.java [hunk 7]
```diff
@@ -60,6 +60,7 @@
   String PROJECTION_OF_POWER = "Projection of Power";
   String ALL_ROCKETS_ATTACK = "All Rockets Attack";
   String ROCKETS_CAN_FLY_OVER_IMPASSABLES = "Rockets Can Fly Over Impassables";
+  String TARGET_ROCKETS_SEQUENTIALLY_AND_AFTER_SBR = "Target Rockets Sequentialy and After SBR";
```
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/Constants.java]: Ok. So all other non-global A&A versions would have them fire before SBR then?

File: game-core/src/main/java/games/strategy/triplea/Constants.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/Constants.java]: The property was used in the code: https://github.com/triplea-game/triplea/pull/4324/files#diff-547357ca4f4a77daa0b2582028381c95L91

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Isn't it actually firing the rockets if the new property is set at this line: https://github.com/triplea-game/triplea/pull/4324/files#diff-547357ca4f4a77daa0b2582028381c95R184

File: game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java]: Given that you moved the v2/v3 calls to battle delegate, I think the last if block should be able to be removed then?

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 8]
```diff
@@ -68,64 +120,93 @@ private static boolean isLimitRocketDamageToProduction(final GameData data) {
...
-  }
-
-  private static void fireWW2V2(
-      final IDelegateBridge bridge,
-      final PlayerID player,
-      final Set<Territory> rocketTerritories) {
+  /**
+   *  Find Rocket Targets and load up fire rockets for later execution if necessary. Directly fired with strict rockets.
+   */
+  private void findRocketTargets(final IDelegateBridge bridge) {
```
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Isn't it actually firing the rockets if the new property is set at this line: https://github.com/triplea-game/triplea/pull/4324/files#diff-547357ca4f4a77daa0b2582028381c95R184

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Ok. I guess as long as we don't have any maps using that property then I'm ok removing it in favor of the tech field. As long as for v2-v6, we still only allow one attack per territory then I think it should be ok.

**Comment by ron-murhammer** [comment]
@simon33-2 Yeah, it was done as V1 was the first map coded then V2 and V3 properties were added to adjust the rules for those maps. I hate that its done that way and want to move towards having separate properties for each portion of the rules that are different rather V2 or V3 checks. Then the V2 or V3 property would just set the appropriate properties in the background so they act as a kind of shortcut for map makers. And yes we have lots of properties that are named rather poorly which why any new ones we add, I tend to want to name that as best as we can.

File: game-core/src/main/java/games/strategy/triplea/Constants.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/Constants.java]: After some reflection, I decided that the Zen theory of Rockets was the same in v2-v5 as it is in Global. That is, one should not be able to change one's mind about where a rocket should be targeted based on the result of (a) another rocket (b) SBR. This is not explicitly stated in the rulebook although I would argue that it is implied. I suppose if you have a hard on for reading the rules in a perverse way you can still add the property.

File: game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/GameStepPropertiesHelper.java]: Oh I've done more than that. I've removed the whole method.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: You know what, you're quite correct. Even though I wrote all of that I had forgotten about that aspect.

I like the idea of breaking up the method so that it only selects a single rocket but that has two problems. 
(1) You need some kind of Tuple to return back both the territory and the unit - you could just continue with adding to the class members though. So not a show stopper by any means.
(2) You'd need to remember the previously attacked Territories. I guess you could make the previouslyAttackedTerritories variable a class member and remember it that way.

Do you want me to do that?

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Do you like the idea of changing the method names or having the check in the move and battle delegates better?

[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: If its the same check then probably better to keep it here to avoid duplication and rename the method. If it isn't the same checks the probably better to push it out to where this is called.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: I'd prefer to avoid class member variables generally as it tends to make things much harder to read and follow. Might be better to just have 2 different methods: 1 for finding all targets for non-global and 1 for dealing with target, fire, target, fire.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java [hunk 9]
```diff
@@ -189,13 +189,14 @@ public void end() {
...
-    if (GameStepPropertiesHelper.isFireRockets(data)) {
-      if (needToDoRockets && TechTracker.hasRocket(bridge.getPlayerId())) {
-        RocketsFireHelper.fireRockets(bridge, bridge.getPlayerId());
-        needToDoRockets = false;
-      }
+    // WW2V1, fires at end of non combat move.
+    // Other versions fire Rockets in BattleDelegate
+    if (needToDoRockets && GameStepPropertiesHelper.isNonCombatMove(data, true)) {
+      if (TechTracker.hasRocket(bridge.getPlayerId())
+          && !Properties.getWW2V2(data) && !Properties.getAllRocketsAttack(data)) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Do we also need to check that its not v3? Or do all v3+ maps have getAllRocketsAttack set to true?

Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: All the v3 maps that I'm aware of set "All Rockets Attack".

I was going to add referring to the WW2V3 property but then I realised that would propagate the crappy way it was done in the past. So I'm really reluctant to do it that way.

Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Alright. I would think v2 should set that property as well?

Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Just checked, it should but surprisingly it doesn't.

There are 82 map xml's which set WW2V2 without setting "All Rockets Attack". Of these, 50 appear to have a tech step. So we'd need to update a number of maps if we got rid of the WW2V2 check.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: Ultimately, I've gone with your original suggestion here of renaming the method.

A general comment on member variables. I've done a lot of stuff in a language that wouldn't allow you to omit the "this." on member variable access (although it called it "CurObject."). Python is similar. Have to say I much prefer this approach to the Java approach. However, given that the Java approach is to not distinguish between local variable access and member access, wouldn't it have been better to retain the "m_" prefix so that we could easily determine when this was done? Apologies if I've asked this before.

Is your issue with class member variables them not being obvious where they are coming from? In general I know what you mean about longer lived variables reducing readability but there's also issues if you pass around a tonne of variables too.

**Comment by simon33-2** [comment]
I had some trouble with out of memory errors testing this. Should we increase the maximum memory in the build.gradle? Please don't increase the minimum too much or at all because it causes errors on 32bit systems.

FWIW, I used "./gradle run -Dexec.args="-Xmx=1500M" to get around the errors. Seemed to work.

**Comment by ron-murhammer** [comment]
@simon33-2 If you rebased on master then it could be related to @ssoloff recent change for memory #4336. If you are seeing errors, it would be good to report them. The default before that change was 2G which was causing problems for 32-bit windows users.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java (hunk 9 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Ok. We can leave the v2 check then. Just hate that some of these properties are so messy and hard to follow.

**Comment by simon33-2** [comment]
I think the errors were caused by "./gradlew run" not using the TripleA.vmoptions file.

**Comment by ssoloff** [comment]
> I think the errors were caused by "./gradlew run" not using the TripleA.vmoptions file.

Note that not only does Gradle not know anything about _TripleA.vmoptions_, but, even if it did, that file doesn't exist at build time--it is created by the installer.

> Should we increase the maximum memory in the build.gradle?

I don't believe we've ever configured the build scripts to specify how much memory Gradle should use when running the application.  I suspect the Gradle (or JVM) default is based on how much total memory is available on the system.  It may be that your system has a relatively small amount of RAM, and some change (whether in this PR or another) is causing the application to run exceed the default maximum amount of heap space Gradle/JVM is calculating.

Figuring out what's eating up heap space wouldn't be a bad thing, especially if that's going to make it impossible to run TripleA on a 32-bit system.  However, if that's the way things roll, and you need to adjust the default heap size on your machine, I don't think the correct solution would be to to globally update the build scripts, which would affect everyone.

Doing what you did above using the CLI is one way to go.  If you'd prefer not to have to do that on every invocation, you can export the `_JAVA_OPTIONS` environment variable once in your terminal, and all subsequent runs will use those VM args.  For example:

```
$ export _JAVA_OPTIONS="-Xmx1500m -Xms512m"
$ ./gradlew :game-headed:run
Picked up _JAVA_OPTIONS: -Xmx1500m -Xms512m

> Task :game-headed:run
Picked up _JAVA_OPTIONS: -Xmx1500m -Xms512m
```

**Comment by simon33-2** [comment]
@ssoloff my thinking was that it would be using a 1G max which is not good for a few maps. Yes, it's only a 2G system. Didn't seem to be running out of memory though. Is it in fact using a dynamically generated max? So if we updated gradle we could see less memory available on some systems. I see the problem.

**Comment by ssoloff** [comment]
@simon33-2 

> my thinking was that it would be using a 1G max which is not good for a few maps. Yes, it's only a 2G system. Is it in fact using a dynamically generated max?

I believe so.  You can verify what maximum heap size the JVM is choosing by running the appropriate command from [this SO answer](https://stackoverflow.com/a/13871564/3900879).

> So if we updated gradle we could see less memory available on some systems. I see the problem.

Exactly.

**Comment by simon33-2** [comment]
Right, so in the context of this PR, running out of heap space is something of a red herring. And there's no need for change, indeed a change would be regressive.

**Comment by ron-murhammer** [comment]
@simon33-2 Sorry, was focusing on the A&A PBF updates. I'll try to pick up reviewing this in the next few days.

**Comment by simon33-2** [comment]
No problem. Totally agree that PBF is the priority.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 10]
```diff
@@ -68,64 +122,96 @@ private static boolean isLimitRocketDamageToProduction(final GameData data) {
...
-      if (oneAttackPerTerritory) {
-        targets.removeAll(attackedTerritories);
+    final PlayerId player = bridge.getPlayerId();
+    final Map<Territory, Integer> previouslyAttackedTerritories = new LinkedHashMap<>();
+    for (final Territory attackFrom : getTerritoriesWithRockets(data, player)) {
+      final Set<Territory> targets = getTargetsWithinRange(attackFrom, data, player);
+      final int maxAttacks = TechAbilityAttachment.getRocketNumberPerTerritory(player, data);
+      for (final Territory t : previouslyAttackedTerritories.keySet()) {
+        // negative Rocket Number per Territory == unlimited
+        if (maxAttacks < 0 || maxAttacks <= previouslyAttackedTerritories.get(t).intValue()) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: If maxAttacks == -1 then shouldn't we not remove this from targets since its unlimited?

**Comment by ron-murhammer** [comment]
@simon33-2 Looks pretty good now just 1 comment. My only other question is what maps did you test this on? I think we should probably test it on at least v1, v2, v3, and global to make sure those all work as intended.

File: game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java (hunk 10 above)
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java]: You're quite correct! I had maxAttacks set to 200 in Global so I didn't spot this bug.

**Comment by simon33-2** [comment]
Just V1, V2, V3 and Global.

V3 is pretty similar to Global just without the facilities. I'll have to retest later with that bug fixed.

**Comment by ron-murhammer** [comment]
@simon33-2 If you fix that and re-test those maps then we should be good to merge this.

**Comment by simon33-2** [comment]
Sorry about the delay in this one. Discovered another bug and didn't have that much time to work on it. Proved pretty easy.

Should be all good now with one proviso. Hit this bug in testing after using Ctrl-Q to "Leave Game":
> Task :game-headed:run
Nov 23, 2018 6:25:27 PM games.strategy.engine.framework.GameRunner lambda$start$0
SEVERE: 234
java.lang.ArrayIndexOutOfBoundsException: 234
        at java.util.stream.SortedOps$SizedRefSortingSink.accept(SortedOps.java:364)
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1553)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at games.strategy.engine.framework.ui.GameChooserModel.<init>(GameChooserModel.java:49)
        at games.strategy.engine.framework.ui.GameChooserModel.<init>(GameChooserModel.java:45)
        at games.strategy.engine.framework.startup.mc.GameSelectorModel.selectByName(GameSelectorModel.java:259)
        at games.strategy.engine.framework.startup.mc.GameSelectorModel.loadDefaultGameSameThread(GameSelectorModel.java:243)
        at java.lang.Thread.run(Thread.java:748)


Doubt that it was to do with anything I had done but I'm going to report it anyway.

Works on V1, V2, V3 & Global. Infinite rockets work on Global too, now.

Hmm, it seems that one of the tests fails now. Adding a parameter to sumNumbers causes trouble for the test.

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java [hunk 11]
```diff
@@ -189,13 +189,11 @@ public void end() {
...
-    // WW2V2/WW2V3, fires at end of combat move
-    // WW2V1, fires at end of non combat move
-    if (GameStepPropertiesHelper.isFireRockets(data)) {
-      if (needToDoRockets && TechTracker.hasRocket(bridge.getPlayerId())) {
-        RocketsFireHelper.fireRockets(bridge, bridge.getPlayerId());
-        needToDoRockets = false;
-      }
+    // WW2V1, fires at end of non combat move.
+    // Other versions fire Rockets in BattleDelegate
+    if (needToDoRockets && GameStepPropertiesHelper.isNonCombatMove(data, true)) {
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: Based on Panther's overview of rockets, I don't think this is correct. V1 should fire at the same time as all the others.

File: game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java [hunk 12]
```diff
@@ -135,11 +135,13 @@ static int sumIntegerMap(final Function<TechAbilityAttachment, IntegerMap<UnitTy
   @VisibleForTesting
   static int sumNumbers(
       final ToIntFunction<TechAbilityAttachment> mapper,
+      final String attachmentType,
       final PlayerId player,
       final GameData data) {
     return TechTracker.getCurrentTechAdvances(player, data).stream()
         .map(TechAbilityAttachment::get)
         .filter(Objects::nonNull)
+        .filter(i -> i.getAttachedTo().toString().equals(attachmentType))
```
Comment (ron-murhammer) [inline game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java]: Can you explain why this filter is needed?

File: game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java (hunk 11 above)
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/delegate/MoveDelegate.java]: No, it isn't completely correct. I was intending that changing V1 rockets would be out of scope for this PR.

Getting V1 rockets working completely correctly is a huge amount of work! It requires allowing an AA Gun to be nominated as a rocket, moving on Combat Move but only to friendly territories, perhaps some other things but that's enough to be in my too hard basket.

File: game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java (hunk 12 above)
Comment (simon33-2) [reply game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java]: Took me a minute to think of it. The problem is that other TechAbilityAttachments for other techs get summed up which only really works if the value is zero.

I'm wondering why a nation with multiple techs doesn't have a rocket range of at least 6.

**Comment by simon33-2** [comment]
Thanks for merging this one @ron-murhammer
